Allow easily renaming multiple nodes

KoBeWi opened the pull request:
I wanted this for a long time. The batch rename dialog is an overkill for such basic operation 😒
![](https://chat.godotengine.org/file-upload/ZRSjKwSLmfvKjTYMe/godot.windows.editor.dev.x86_64_TCKSVKFQJ5.gif)
Putting as draft, because undo is broken.

KoBeWi commented:
Undo error fixed, but I had to do a *little* refactoring.

AttackButton commented:
Is it possible to rename an interleaved selection? Like:

node2, node6, node8, node9 -> Test, Test2, Test3, Test4?

KoBeWi commented:
![godot windows editor dev x86_64_0dq1RCl2ak](https://user-images.githubusercontent.com/2223172/209372446-75cdf624-76f6-4a19-ab4c-c42dd4801985.gif)

ajreckof commented:
So I took a look at it, the code seems great and straight forward, I don't have any problem with the implementation. But while doing the worst i could to make it break I did found a few bug and strange behaviour I'll try to order them from most important to sidenotes that are not really important and might only be a personnal point of view.

1. When editing multiple node. StringPath will be badly updated which isn't a case when editing element one at a time. But this also happens with Batch edit so maybe it is okay to have a bug with this here too. (I'm not sure it has been reported yet for Batch Edit). it feels like the rename function is not designed to have multiple rename in the same frame
2. You can edit outside the list by going to a node with the arrows. This is not a problem in general but will make strange behavior especially since there is not enough readability. For example, if you select a node you rename it you then go down with the arrow and rename this other node and unfortunately the previous node is also renamed at the same time. I think if the node is outside of the selection it should not rename selected nodes. 
3. When changing the name of one node to an empty string it will rename the node to the name of its class but this behavior was not reimplemented in multi edit so it will just rename everyone from the name of the class being edited instead of their own class.
4. Edit disappear from popup when multiple node are selected since we are adding a particular behavior on this it might be a good idea to make it reappear maybe with a slightly modified name. Simple Batch Edit? Multi-edit ?
5. After multi editing the inspector will edit the first node instead of the batch of node
6. When modifying the name to the name of the focused node it won't rename any of the node as the name of the focused node didn't change. It might be problematic if someone is editing the property of a batch of node and then want to rename the nodes and just paste a value and nothing happens. Right now you have to either unselect the node modify and then reselect your node to continue with your edition of those nodes or set the value to something arbitrary and then setting to the value you want. 
6. Right now it feels like we are only editing the focused node. I’m not sure how to improve it :(

KoBeWi commented:
1. What is StringPath 🤔
2. See this issue: #12544 If you move within nodes of the same type, the nodes are correctly deselected
3. Makes sense
4. Makes sense
5. The inspector is consistent with what is selected after rename, so it's not really a problem
6. I'll see if it can be fixed easily
7. The PR was inspired by Windows' file explorer, which also behaves like this
![image](https://user-images.githubusercontent.com/2223172/232100897-7d03c860-baa1-4dd5-9eba-32cf4ca507ef.png)

ajreckof commented:
1. sorry for first one it is NodePath. I will make a different bug issue. I think the solution does not need to be in this PR as the problem is also present in Batch Edit and most likely the solution should be in both. I'll look if the issue already exists. If it does not I will create it. If it does I will update this with a link to it.
2. okay so it was just another bug
3. N/A
4. N/A
5. N/A
6. N/A
7.  That's why i placed it last as i didn't have any idea how to make it better and didn't have any other example of similar behaviour, if you feel it is easily understandable by most people than i think it's okay

KoBeWi committed:
Allow to easily rename multiple nodes

Co-authored-by: ajreckof <tbonhoure@ymail.Com>

KoBeWi commented:
Rebased after #76376
I reworked the code using the new methods and it's now simpler.

I tested different cases and it works correctly, but seems to break node paths in very specific cases .-. Maybe it's something that could be fixed later, idk.

ajreckof commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1012,32 +1017,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
+	}
+
 	if (new_name.strip_edges().is_empty()) {
 		// If name is empty, fallback to class name.
 		if (GLOBAL_GET("editor/naming/node_name_casing").operator int() != NAME_CASING_PASCAL_CASE) {
 			new_name = Node::adjust_name_casing(n->get_class());
 		} else {
 			new_name = n->get_class();
 		}
+
+		// When doing batch rename, rename all nodes to their respective class.
+		if (p_batch_item == p_item) {
+			p_batch_item->set_meta("use_class", true);
+		}
 	}
 
 	if (n->is_unique_name_in_owner() && get_tree()->get_edited_scene_root()->get_node_or_null("%" + new_name) != nullptr) {
 		error->set_text(TTR("Another node already uses this unique name in the scene."));
 		error->popup_centered();
-		which->set_text(0, n->get_name());
+		p_item->set_text(0, n->get_name());
 		return;
 	}
```
While looking at this I realized it was a problem they weren't inside `rename_node` so I tried a bit to see if I could things broke on master because of those and effectively I can :( So opening an issue for each ( #78290 
#78291) and a fix PR #78292

ajreckof commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1012,32 +1017,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
+	}
+}
+
+void SceneTreeEditor::_renamed(TreeItem *p_item, TreeItem *p_batch_item, Node *p_node) {
+	Node *n;
+	if (p_node) {
+		n = p_node; // During batch rename the paths may change, so using metadata is unreliable.
+	} else {
+		n = get_node(p_item->get_metadata(0));
+	}
+	ERR_FAIL_NULL(n);
+
+	String new_name;
+	if (p_batch_item) {
+		if (!p_batch_item->get_meta(SNAME("use_class"), false)) {
+			new_name = p_batch_item->get_text(0);
+		}
+	} else {
+		new_name = p_item->get_text(0);
+	}
```
once the lower part is handled by `rename_node` (#78292).I think the rest of this function can be integrated into the `_edited` function this will make clearer and more readable code.

ajreckof reviewed (changes requested):
So reviewing your code made me see some short comings of #76376 I put some review but most likely this PR will be better after I fixed the shortcomings I found. 

> I tested different cases and it works correctly, but seems to break node paths in very specific cases .-. Maybe it's something that could be fixed later, idk.

Most likely this is this issue #76192. There is a PR but it has to be rebased on my last changes on this part. 
Edit : you found it while I was reviewing ^__^

KoBeWi commented:
@ajreckof Would you be interested in doing the changes you requested? I rebased and tested the PR and it still works. If I understand correctly, it *might* have some inconsistencies with batch rename, but right now it's in mergable state and this can be improved in a follow-up.

KoBeWi commented:
Which bugs?

ajreckof commented:
My bad I misremembered the suggestions I don't know why :( forget what I said. Modifications are even easier.

Mickeon commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1108,17 +1113,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
+	}
+
+	if (n->is_unique_name_in_owner() && get_tree()->get_edited_scene_root()->get_node_or_null("%" + new_name) != nullptr) {
+		error->set_text(TTR("Another node already uses this unique name in the scene."));
```
Mm...
```suggestion
		error->set_text(vformat(TTR("A node with the unique name %s already exists in this scene."), new_name));
```

Mickeon commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1108,17 +1113,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
+
+	if (new_name.strip_edges().is_empty()) {
+		// If name is empty, fallback to class name.
+		if (GLOBAL_GET("editor/naming/node_name_casing").operator int() != NAME_CASING_PASCAL_CASE) {
```
Is this preferable over `int(GLOBAL_GET("editor/naming/node_name_casing"))` or is it just a codestyle choice?

KoBeWi replied:
Personal choice.
Though nowadays I think I prefer `int()` syntax, idk.

Mickeon reviewed (commented):
This looks great, not tested though

ajreckof commented:
here are the modifications I suggested on top of the last commit you pushed : https://github.com/KoBeWi/godot/compare/million_of_names...ajreckof:godot:millions-of-names?expand=1

KoBeWi commented:
I actually tried doing your reverse list suggestion (before I wrote my comment), and it led me to the same result:

https://github.com/godotengine/godot/assets/2223172/504ce0de-9dc6-4a40-adac-b3cd8a8edef3

The nodes are named in reverse. That's not an expected behavior.

ajreckof commented:
I know it will be named in reverse see https://github.com/godotengine/godot/pull/69087#discussion_r1231515895 the fact is this is already the case with batched rename so I think it would be better to first make it like that so that it is consistent and create a second PR that will fix the order of the renames in both cases. If you really wish to add it here I would then also fix the batch rename side too.

KoBeWi commented:
I'd rather make the naming correct and your changes + related fixes can be done in a follow up.

AThousandShips commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1030,8 +1030,13 @@ void SceneTreeEditor::set_selected(Node *p_node, bool p_emit_selected) {
-void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
-	TreeItem *item = _find(tree->get_root(), p_node->get_path());
+void SceneTreeEditor::rename_node(Node *p_node, const String &p_name, TreeItem *p_item) {
+	TreeItem *item;
```
```suggestion
	TreeItem *item = nullptr;
```

KoBeWi replied:
This is intentional. The variable is initialized in the following block.

AThousandShips commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1108,17 +1113,74 @@ void SceneTreeEditor::_rename_node(Node *p_node, const String &p_name) {
+}
+
+void SceneTreeEditor::_renamed(TreeItem *p_item, TreeItem *p_batch_item, Node *p_node) {
+	Node *n;
```
```suggestion
	Node *n = nullptr;
```

KoBeWi replied:
Same.

Mickeon reviewed (approved):
I feel like trusting @KoBeWi on this one (not tested).

Code looks good and thank the heavens, the feature itself is nice and reminiscent of other software.

ajreckof commented:
> I'd rather make the naming correct and your changes + related fixes can be done in a follow up.

That seems okay I would still include my suggestion of removing the _renamed function and calling directly the rename_node function with the right parameter as what is done in _renamed is already done in rename node

KoBeWi commented:
Uh can you make a separate patch for that? >_>

ajreckof commented:
of course here is the updated patch https://github.com/KoBeWi/godot/compare/million_of_names...ajreckof:godot:millions-of-names?expand=1

AThousandShips commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -1491,7 +1514,7 @@ void SceneTreeEditor::set_connecting_signal(bool p_enable) {
 void SceneTreeEditor::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("_update_tree"), &SceneTreeEditor::_update_tree, DEFVAL(false)); // Still used by UndoRedo.
-	ClassDB::bind_method("_rename_node", &SceneTreeEditor::_rename_node);
+	ClassDB::bind_method("_test_update_tree", &SceneTreeEditor::_test_update_tree);
```
This isn't used anywhere, any reason it's bound?

nongvantinh commented:
I think this feature could coexist with the existing batch rename dialog #15928. However, from a UX perspective, it would be beneficial if, instead of just displaying one `LineEdit` for the first selected node, we could display `LineEdits` for all selected nodes and update changes as we type.

Mickeon commented:
> we could display LineEdits for all selected nodes and update changes as we type

Indeed, it's not possible without some nasty hacks and workarounds. Not really an issue in my opinion, but it may hurt discoverability.

ajreckof commented:
> > we could display LineEdits for all selected nodes and update changes as we type
> 
> Indeed, it's not possible without some nasty hacks and workarounds. Not really an issue in my opinion, but it may hurt discoverability.
> 
> For this to be more discoverable both **Rename** and **Batch Rename** should be available on the popup menu when selecting multiple nodes but that remains to be seen later.

strange see how I proposed it here https://github.com/godotengine/godot/pull/69087#issuecomment-1508856354 and it was said to be resolved here https://github.com/godotengine/godot/pull/69087#issuecomment-1509010046 but effectively when I test it isn't appearing



> I think this feature could coexist with the existing batch rename dialog #15928. However, from a UX perspective, it would be beneficial if, instead of just displaying one `LineEdit` for the first selected node, we could display `LineEdits` for all selected nodes and update changes as we type. However, the limitation of the current system only allows editing one node at a time. Therefore, I believe it is impossible to achieve.

see for better context : https://github.com/godotengine/godot/pull/69087#issuecomment-1508918621

KoBeWi commented:
I fixed the missing Rename option for multiple nodes.

Calinou reviewed (commented):
Tested locally (rebased on top of `master` f2045ba822bff7d34964901393581a3117c394a9), it mostly works as expected.

Only entries that are located *below* the currently focused TreeItem (see the gray focus outline) are renamed, so if you select from bottom to top, only one entry will be renamed.

https://github.com/godotengine/godot/assets/180032/a9e11ebe-299a-4677-841f-fc9cfd02ed0f

Considering there's been talk to hide this gray focus outline in Tree, I suggest not changing the behavior based on where the focused item is and always rename all selected items.

KoBeWi commented:
Fixed.

ajreckof commented:
When the root is selected it will not rename it and start from the next one.

KoBeWi commented:
https://github.com/godotengine/godot/pull/69087#issuecomment-1974940415

Calinou reviewed (approved):
Tested locally, it works as expected now (including when nodes at different levels are selected, or when both parent and children are selected).

akien-mga commented:
Thanks!
